app/gui/src/dashboard/components/AriaComponents/Form/components/useForm.ts

```diff
@@ -59,190 +70,248 @@ export function useForm<Schema extends types.TSchema, SubmitResult = void>(
-
-    const formInstance = reactHookForm.useForm({
-      ...options,
-      resolver: zodResolver.zodResolver(
-        computedSchema,
-        {
-          async: true,
-          errorMap: (issue) => {
-            switch (issue.code) {
-              case 'too_small':
-                if (issue.minimum === 1 && issue.type === 'string') {
-                  return {
-                    message: getText('arbitraryFieldRequired'),
-                  }
-                } else {
-                  return {
-                    message: getText('arbitraryFieldTooSmall', issue.minimum.toString()),
-                  }
-                }
-              case 'too_big':
-                return {
-                  message: getText('arbitraryFieldTooLarge', issue.maximum.toString()),
-                }
-              case 'invalid_type':
+  /* eslint-disable react-compiler/react-compiler, react-hooks/rules-of-hooks */
+  if (isFormInstance(optionsOrFormInstance)) {
+    return useFormInstance(optionsOrFormInstance, formInstanceCallbacks)
+  }
+
+  return useCreateFormProps(optionsOrFormInstance)
+  /* eslint-enable react-hooks/rules-of-hooks, react-compiler/react-compiler */
+}
+
+/**
+ * Checks if the argument is a form instance
+ */
+function isFormInstance<Schema extends types.TSchema, SubmitResult>(
+  args: types.UseFormOptions<Schema, SubmitResult> | types.UseFormReturn<Schema, SubmitResult>,
+): args is types.UseFormReturn<Schema, SubmitResult> {
+  return 'formState' in args
+}
+
+/** Get the type of arguments passed to the useForm hook */
+function getArgsType<Schema extends types.TSchema, SubmitResult = void>(
+  args: types.UseFormOptions<Schema, SubmitResult>,
+): 'formInstance' | 'formOptions' {
+  return isFormInstance(args) ? 'formInstance' : 'formOptions'
+}
+
+/**
+ * Returns Form instance with merged callbacks
+ * @internal
+ */
+function useFormInstance<Schema extends types.TSchema, SubmitResult>(
+  form: types.UseFormReturn<Schema, SubmitResult>,
+  formInstanceCallbacks: types.OnSubmitCallbacks<Schema, SubmitResult> = {},
+): types.UseFormReturn<Schema, SubmitResult> {
+  // mutations here are necessary, because we want to mutate the form instance,
+  // instead of creating a new one, to make callbacks work
+  const props = mergeProps<types.OnSubmitCallbacks<Schema, SubmitResult>>()(
+    {
+      onSubmit: form.onSubmit,
+      onSubmitSuccess: form.onSubmitSuccess,
+      onSubmitFailed: form.onSubmitFailed,
+      onSubmitted: form.onSubmitted,
+    },
+    {
+      onSubmit: formInstanceCallbacks.onSubmit,
+      onSubmitSuccess: formInstanceCallbacks.onSubmitSuccess,
+      onSubmitFailed: formInstanceCallbacks.onSubmitFailed,
+      onSubmitted: formInstanceCallbacks.onSubmitted,
+    },
+  )
+
+  // @ts-expect-error - This is safe, because even though the property is readonly,
+  // it's readonly only for external usage, and we're merging the callbacks internally
+  form.onSubmit = props.onSubmit
```

Comment:
we can use `unsafeMutable` here! but should still keep the comment for explanation

Reply:
Thanks!